New push frame stitch application

jessemapel opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

Adds a new application for stitching push frame images back together.

## Related Issue
<!--- This project only accepts pull requests related to open issues (GitIssues) -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

resolves #4924 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

We need this application for processing push frame imagery such as LRO WAC outside of ISIS. This will allow us to ingest the images into ISIS and do processing such as radiometric calibration and then export a single image with all of the image data in it. It will also allow us to flip the frame so that adjacent ground points are adjacent in the image and we can use matchers without fear of ending up on frame boundaries.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested on an LRO WAC UV and VIS image set. There is also extensive automated testing.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [x] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

jessemapel committed:
Added new framestitch app

jessemapel committed:
Added changelog entry

oleg-alexandrov commented on CHANGELOG.md:
```diff
@@ -35,6 +35,8 @@ release.
 
 ## [Unreleased]
 - Updated the LRO photometry application Lronacpho, to use by default the current 2019 photometric model (LROC_Empirical). The model's coefficients are found in the PVL file that exists in the LRO data/mission/calibration directory. If the old parameter file is provided, the old algorithm(2014) will be used. This functionality is desired for calculation comparisons. Issue: [#4512](https://github. ...
+- Added new application, framestitch, for stitching even and odd push frame images back together prior to processing in other applications. [4924](https://github.com/USGS-Astrogeology/ISIS3/issues/4924)
```
Should be "a new application".

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+namespace Isis {
+
+  /**
+   * Helper function to count the size of frames in a cube
```
It looks that the comment is not aligned with what the function does. The function appears to find the median framelet size rather than counting. The function name may also need to be renamed to computeMedianFrameSize() or computeMedianFrameletSize(), and somewhere it should be explained what a framelet is.

I think this computes in fact framelet height, rather than its 2D size.

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+  }
+
+  /**
+   * Compute the stats for an ISIS cube. This is the programmatic interface to
```
This comment also appears to not say what is actually happening. This function does the stitching, rather than the stats.

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+      int oddFrameSize = computeFrameSize(oddCube);
+
+      if (evenFrameSize == 0) {
+        QString msg = "Failed to find any NULL frames in even cube when attempting "
```
I think this should error message should say "non-NULL" rather than "NULL". Or even better, maybe, "failed to find any valid data in even cube".

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      if (oddFrameSize == 0) {
+        QString msg = "Failed to find any NULL frames in odd cube when attempting "
```
Same thing here.

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+    // 0000  ####         ####  0000
+    // ####  0000         0000  ####
+    //
+    int numFrames = evenCube->lineCount() / frameSize;
```
I think this implicitly assumes that all frames have the same number of lines (frameSize). I am also a little confused here. I would think that lineCount() includes both the lines which have nulls (which look black) and the lines which have valid data, no?

jessemapel replied:
Yes this is true. All frames will have the same number of lines. We want to include all of the lines, including the NULL frames, because we will need to iterate through every frame.

oleg-alexandrov replied:
I see. So then the frame size includes both valid and invalid lines. I think that was not clear from the comment on top of that function. I assumed only valid lines are counted.

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,220 @@
+                      "to compute frame size.";
+        throw IException(IException::User, msg, _FILEINFO_);
+      }
+      if (evenFrameSize != oddFrameSize) {
```
This implicitly assumes that all frames in both even and odd cubes have the same size. If they don't, given that the frame size is found using a median, these two medians will likely disagree.

jessemapel replied:
You can't stitch together two images if you have two different frame sizes as the frame boundaries won't line up.

oleg-alexandrov replied:
Then finding the median is not helpful as all values should be equal. Instead, when, for a given even or odd cube the height of each frame is found, it should be verified right there that all obtained frames in a cube have the same height, before comparing the frame heights between the cubes.

jessemapel replied:
I went with the median to account for any weird situations where there's random lines of NULL in the image that are not associated with missing even/odd frames. I could see just validating in the counting function being better. I'll switch to that. If there's a situation where there are random all NULL lines, then the user can just manually enter the frame size.

oleg-alexandrov replied:
Sounds good. That was more for my own understanding.

oleg-alexandrov commented on isis/src/base/apps/framestitch/framestitch.xml:
```diff
@@ -0,0 +1,124 @@
+      This program combines the even and odd frames from push frame <def link="Cube">cubes</def> back
+      into a single interwoven <def link="Cube">cube</def>. It can also optionally flip the frame order
+      so that adjacent ground features appear adjacent between frames. This program
+      is designed to be ran before processing the data in another software package
```
should be "to be run".

oleg-alexandrov commented on isis/src/base/objs/ProcessByBrick/ProcessByBrick.cpp:
```diff
@@ -660,6 +660,82 @@ void ProcessByBrick::SetOutputRequirements(int outputRequirements) {
 
+
+  /**
+   * Starts the systematic processing of the input cube by moving an arbitrary
```
should be "arbitrarily-shaped".

jessemapel committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into push_frame_stitch

jessemapel committed:
Code updates

jessemapel commented:
Updated the docs, re-worked the error handling in the automatic frame height computations, and replaced all instances of frame size with frame height to be more clear.

AustinSanders commented on isis/src/base/apps/framestitch/framestitch.cpp:
```diff
@@ -0,0 +1,221 @@
+    }
+
+    // Use the median to handle any outliers
+    return frameHeights.Average();
```
Should this be average or median?

jessemapel replied:
I switched to a statistics object instead of a histogram and forgot to update the comment. It doesn't matter because there should only be 1 value accumulated by the time we get here

AustinSanders commented on isis/src/base/objs/ProcessByBrick/ProcessByBrick.cpp:
```diff
@@ -660,6 +660,82 @@ void ProcessByBrick::SetOutputRequirements(int outputRequirements) {
+   * brick through the cube. This method allows multiple input and output
+   * cubes.
+   *
+   * @deprecated Please use ProcessCubeInPlace, ProcessCube, or ProcessCubes
```
Are we adding deprecated functionality here?

jessemapel replied:
These deprecated tags need to be removed, the functional signatures are brand new.

jessemapel committed:
Fixed comment, removed dep tags

oleg-alexandrov reviewed (approved):
I have no further comments. Whenever the other reviewer approves this can be pulled in.
